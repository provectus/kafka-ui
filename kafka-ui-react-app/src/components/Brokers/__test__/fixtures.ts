import { BrokersLogdirs } from 'generated-sources';

export const brokersPayload = [
  { id: 1, host: 'b-1.test.kafka.amazonaws.com', port: 9092 },
  { id: 2, host: 'b-2.test.kafka.amazonaws.com', port: 9092 },
];

export const clusterStatsPayload = {
  brokerCount: 2,
  activeControllers: 1,
  onlinePartitionCount: 138,
  offlinePartitionCount: 0,
  inSyncReplicasCount: 239,
  outOfSyncReplicasCount: 0,
  underReplicatedPartitionCount: 0,
  diskUsage: [
    { brokerId: 0, segmentSize: 334567, segmentCount: 245 },
    { brokerId: 1, segmentSize: 12345678, segmentCount: 121 },
  ],
  version: '2.2.1',
};

const partition = {
  broker: 2,
  offsetLag: 0,
  partition: 2,
  size: 0,
};
const topics = {
  name: '_confluent-ksql-devquery_CTAS_NUMBER_OF_TESTS_59-Aggregate-Aggregate-Materialize-changelog',
  partitions: [partition],
};

export const brokerLogDirsPayload: BrokersLogdirs[] = [
  {
    error: 'NONE',
    name: '/opt/kafka/data-0/logs',
    topics: [
      {
        ...topics,
        partitions: [partition, partition, partition],
      },
      topics,
      {
        ...topics,
        partitions: [],
      },
    ],
  },
];
