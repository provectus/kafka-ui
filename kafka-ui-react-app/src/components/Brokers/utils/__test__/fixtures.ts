import { BrokerMetrics } from 'generated-sources';

export const brokerMetricsPayload: BrokerMetrics = {
  segmentSize: 23,
  segmentCount: 23,
  metrics: [
    {
      name: 'TotalFetchRequestsPerSec',
      labels: {
        canonicalName:
          'kafka.server:name=TotalFetchRequestsPerSec,topic=_connect_status,type=BrokerTopicMetrics',
      },
      value: 10,
    },
    {
      name: 'ZooKeeperRequestLatencyMs',
      value: 11,
    },
    {
      name: 'RequestHandlerAvgIdlePercent',
    },
  ],
};
export const transformedBrokerMetricsPayload =
  '{"segmentSize":23,"segmentCount":23,"metrics":[{"name":"TotalFetchRequestsPerSec","labels":{"canonicalName":"kafka.server:name=TotalFetchRequestsPerSec,topic=_connect_status,type=BrokerTopicMetrics"},"value":10},{"name":"ZooKeeperRequestLatencyMs","value":11},{"name":"RequestHandlerAvgIdlePercent"}]}';
