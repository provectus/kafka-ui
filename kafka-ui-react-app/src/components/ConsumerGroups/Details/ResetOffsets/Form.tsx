import React from 'react';
import { useNavigate } from 'react-router-dom';
import {
  ConsumerGroupDetails,
  ConsumerGroupOffsetsReset,
  ConsumerGroupOffsetsResetType,
} from 'generated-sources';
import { ClusterGroupParam } from 'lib/paths';
import {
  Controller,
  FormProvider,
  useFieldArray,
  useForm,
} from 'react-hook-form';
import { MultiSelect, Option } from 'react-multi-select-component';
import 'react-datepicker/dist/react-datepicker.css';
import { ErrorMessage } from '@hookform/error-message';
import { InputLabel } from 'components/common/Input/InputLabel.styled';
import { Button } from 'components/common/Button/Button';
import Input from 'components/common/Input/Input';
import { FormError } from 'components/common/Input/Input.styled';
import useAppParams from 'lib/hooks/useAppParams';
import { useResetConsumerGroupOffsetsMutation } from 'lib/hooks/api/consumers';
import { FlexFieldset, StyledForm } from 'components/common/Form/Form.styled';
import ControlledSelect from 'components/common/Select/ControlledSelect';

import * as S from './ResetOffsets.styled';

interface FormProps {
  defaultValues: ConsumerGroupOffsetsReset;
  topics: string[];
  partitions: ConsumerGroupDetails['partitions'];
}

const resetTypeOptions = Object.values(ConsumerGroupOffsetsResetType).map(
  (value) => ({ value, label: value })
);

const Form: React.FC<FormProps> = ({ defaultValues, partitions, topics }) => {
  const navigate = useNavigate();
  const routerParams = useAppParams<ClusterGroupParam>();
  const reset = useResetConsumerGroupOffsetsMutation(routerParams);
  const topicOptions = React.useMemo(
    () => topics.map((value) => ({ value, label: value })),
    [topics]
  );
  const methods = useForm<ConsumerGroupOffsetsReset>({
    mode: 'onChange',
    defaultValues,
  });

  const {
    handleSubmit,
    setValue,
    watch,
    control,
    formState: { errors },
  } = methods;
  const { fields } = useFieldArray({
    control,
    name: 'partitionsOffsets',
  });

  const resetTypeValue = watch('resetType');
  const topicValue = watch('topic');
  const offsetsValue = watch('partitionsOffsets');
  const partitionsValue = watch('partitions') || [];

  const partitionOptions =
    partitions
      ?.filter((p) => p.topic === topicValue)
      .map((p) => ({
        label: `Partition #${p.partition.toString()}`,
        value: p.partition,
      })) || [];

  const onSelectedPartitionsChange = (selected: Option[]) => {
    setValue(
      'partitions',
      selected.map(({ value }) => value)
    );

    setValue(
      'partitionsOffsets',
      selected.map(({ value }) => {
        const currentOffset = offsetsValue?.find(
          ({ partition }) => partition === value
        );
        return { offset: currentOffset?.offset, partition: value };
      })
    );
  };

  React.useEffect(() => {
    onSelectedPartitionsChange([]);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [topicValue]);

  const onSubmit = async (data: ConsumerGroupOffsetsReset) => {
    await reset.mutateAsync(data);
    navigate('../');
  };

  return (
    <FormProvider {...methods}>
      <StyledForm onSubmit={handleSubmit(onSubmit)}>
        <FlexFieldset>
          <ControlledSelect
            name="topic"
            label="Topic"
            placeholder="Select Topic"
            options={topicOptions}
          />
          <ControlledSelect
            name="resetType"
            label="Reset Type"
            placeholder="Select Reset Type"
            options={resetTypeOptions}
          />
          <div>
            <InputLabel>Partitions</InputLabel>
            <MultiSelect
              options={partitionOptions}
              value={partitionsValue.map((p) => ({
                value: p,
                label: String(p),
              }))}
              onChange={onSelectedPartitionsChange}
              labelledBy="Select partitions"
            />
          </div>
          {resetTypeValue === ConsumerGroupOffsetsResetType.TIMESTAMP &&
            partitionsValue.length > 0 && (
              <div>
                <InputLabel>Timestamp</InputLabel>
                <Controller
                  control={control}
                  name="resetToTimestamp"
                  rules={{
                    required: 'Timestamp is required',
                  }}
                  render={({ field: { onChange, onBlur, value, ref } }) => (
                    <S.DatePickerInput
                      ref={ref}
                      selected={new Date(value as number)}
                      onChange={(e: Date | null) => onChange(e?.getTime())}
                      onBlur={onBlur}
                    />
                  )}
                />
                <ErrorMessage
                  errors={errors}
                  name="resetToTimestamp"
                  render={({ message }) => <FormError>{message}</FormError>}
                />
              </div>
            )}

          {resetTypeValue === ConsumerGroupOffsetsResetType.OFFSET &&
            partitionsValue.length > 0 && (
              <S.OffsetsWrapper>
                {fields.map((field, index) => (
                  <Input
                    key={field.id}
                    label={`Partition #${field.partition} Offset`}
                    type="number"
                    name={`partitionsOffsets.${index}.offset` as const}
                    hookFormOptions={{
                      shouldUnregister: true,
                      required: 'Offset is required',
                      min: {
                        value: 0,
                        message: 'must be greater than or equal to 0',
                      },
                    }}
                    withError
                  />
                ))}
              </S.OffsetsWrapper>
            )}
        </FlexFieldset>
        <div>
          <Button
            buttonSize="M"
            buttonType="primary"
            type="submit"
            disabled={partitionsValue.length === 0}
          >
            Submit
          </Button>
        </div>
      </StyledForm>
    </FormProvider>
  );
};

export default Form;
