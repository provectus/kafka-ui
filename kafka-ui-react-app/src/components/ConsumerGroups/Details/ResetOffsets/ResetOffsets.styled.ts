import styled from 'styled-components';
import DatePicker from 'react-datepicker';

export const OffsetsWrapper = styled.div`
  display: flex;
  width: 100%;
  flex-wrap: wrap;
  gap: 16px;
`;

export const DatePickerInput = styled(DatePicker).attrs({
  showTimeInput: true,
  timeInputLabel: 'Time:',
  dateFormat: 'MMMM d, yyyy h:mm aa',
})`
  height: 40px;
  border: 1px ${({ theme }) => theme.select.borderColor.normal} solid;
  border-radius: 4px;
  font-size: 14px;
  width: 270px;
  padding-left: 12px;
  background-color: ${({ theme }) => theme.input.backgroundColor.normal};
  color: ${({ theme }) => theme.input.color.normal};
  &::placeholder {
    color: ${({ theme }) => theme.input.color.normal};
  }
  &:hover {
    cursor: pointer;
  }
  &:focus {
    outline: none;
  }
`;
