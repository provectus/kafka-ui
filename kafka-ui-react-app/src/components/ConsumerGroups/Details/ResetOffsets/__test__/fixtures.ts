export const expectedOutputs = {
  EARLIEST: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: undefined,
        partition: 0,
      },
    ],
    resetType: 'EARLIEST',
    topic: '__amazon_msk_canary',
  },
  LATEST: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: undefined,
        partition: 0,
      },
    ],
    resetType: 'LATEST',
    topic: '__amazon_msk_canary',
  },
  OFFSET: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: '10',
        partition: 0,
      },
    ],
    resetType: 'OFFSET',
    topic: '__amazon_msk_canary',
  },
};
