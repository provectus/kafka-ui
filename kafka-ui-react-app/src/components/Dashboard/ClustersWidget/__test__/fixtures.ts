import { Cluster, ServerStatus } from 'generated-sources';

export const onlineCluster: Cluster = {
  name: 'secondLocal',
  defaultCluster: false,
  status: ServerStatus.Online,
  brokerCount: 1,
  onlinePartitionCount: 6,
  topicCount: 3,
  bytesInPerSec: 0.000030618196853764715,
  bytesOutPerSec: 5.737800890036267075817,
};

export const offlineCluster: Cluster = {
  name: 'local',
  defaultCluster: true,
  status: ServerStatus.Offline,
  brokerCount: 1,
  onlinePartitionCount: 2,
  topicCount: 2,
  bytesInPerSec: 8000.0000067376808542600021,
  bytesOutPerSec: 0.8153063567297119490871,
};

export const clusters: Cluster[] = [onlineCluster, offlineCluster];
