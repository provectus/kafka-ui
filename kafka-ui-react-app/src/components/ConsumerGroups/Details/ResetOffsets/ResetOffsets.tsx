import React from 'react';
import { clusterConsumerGroupsPath, ClusterGroupParam } from 'lib/paths';
import 'react-datepicker/dist/react-datepicker.css';
import PageHeading from 'components/common/PageHeading/PageHeading';
import useAppParams from 'lib/hooks/useAppParams';
import { useConsumerGroupDetails } from 'lib/hooks/api/consumers';
import PageLoader from 'components/common/PageLoader/PageLoader';
import {
  ConsumerGroupOffsetsReset,
  ConsumerGroupOffsetsResetType,
} from 'generated-sources';

import Form from './Form';

const ResetOffsets: React.FC = () => {
  const routerParams = useAppParams<ClusterGroupParam>();

  const consumerGroup = useConsumerGroupDetails(routerParams);

  if (consumerGroup.isLoading || !consumerGroup.isSuccess)
    return <PageLoader />;

  const partitions = consumerGroup.data.partitions || [];
  const { topic } = partitions[0];

  const uniqTopics = Array.from(
    new Set(partitions.map((partition) => partition.topic))
  );

  const defaultValues: ConsumerGroupOffsetsReset = {
    resetType: ConsumerGroupOffsetsResetType.EARLIEST,
    topic,
    partitionsOffsets: [],
    resetToTimestamp: new Date().getTime(),
  };

  return (
    <>
      <PageHeading
        text="Reset offsets"
        backTo={clusterConsumerGroupsPath(routerParams.clusterName)}
        backText="Consumers"
      />
      <Form
        defaultValues={defaultValues}
        topics={uniqTopics}
        partitions={partitions}
      />
    </>
  );
};

export default ResetOffsets;
