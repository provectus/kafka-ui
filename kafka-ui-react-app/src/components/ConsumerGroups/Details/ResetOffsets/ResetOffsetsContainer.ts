import { RouteComponentProps, withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { ClusterName, ConsumerGroupID, RootState } from 'redux/interfaces';
import {
  getConsumerGroupByID,
  getIsConsumerGroupDetailsFetched,
  getOffsetReset,
} from 'redux/reducers/consumerGroups/selectors';
import {
  fetchConsumerGroupDetails,
  resetConsumerGroupOffsets,
  resetConsumerGroupOffsetsAction,
} from 'redux/actions';

import ResetOffsets from './ResetOffsets';

interface RouteProps {
  clusterName: ClusterName;
  consumerGroupID: ConsumerGroupID;
}

type OwnProps = RouteComponentProps<RouteProps>;

const mapStateToProps = (
  state: RootState,
  {
    match: {
      params: { consumerGroupID, clusterName },
    },
  }: OwnProps
) => ({
  clusterName,
  consumerGroupID,
  consumerGroup: getConsumerGroupByID(state, consumerGroupID),
  detailsAreFetched: getIsConsumerGroupDetailsFetched(state),
  IsOffsetReset: getOffsetReset(state),
});

const mapDispatchToProps = {
  fetchConsumerGroupDetails,
  resetConsumerGroupOffsets,
  resetResettingStatus: resetConsumerGroupOffsetsAction.cancel,
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(ResetOffsets)
);
