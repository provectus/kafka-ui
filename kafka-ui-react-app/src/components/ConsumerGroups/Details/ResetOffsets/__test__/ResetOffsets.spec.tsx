import React from 'react';
import fetchMock from 'fetch-mock';
import { act, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { render, WithRoute } from 'lib/testHelpers';
import { clusterConsumerGroupResetOffsetsPath } from 'lib/paths';
import { consumerGroupPayload } from 'redux/reducers/consumerGroups/__test__/fixtures';
import ResetOffsets from 'components/ConsumerGroups/Details/ResetOffsets/ResetOffsets';

const clusterName = 'cluster1';
const { groupId } = consumerGroupPayload;

const renderComponent = () =>
  render(
    <WithRoute path={clusterConsumerGroupResetOffsetsPath()}>
      <ResetOffsets />
    </WithRoute>,
    {
      initialEntries: [
        clusterConsumerGroupResetOffsetsPath(
          clusterName,
          consumerGroupPayload.groupId
        ),
      ],
    }
  );

const resetConsumerGroupOffsetsMockCalled = () =>
  expect(
    fetchMock.called(
      `/api/clusters/${clusterName}/consumer-groups/${groupId}/offsets`
    )
  ).toBeTruthy();

const selectresetTypeAndPartitions = async (resetType: string) => {
  userEvent.click(screen.getByLabelText('Reset Type'));
  userEvent.click(screen.getByText(resetType));
  userEvent.click(screen.getByText('Select...'));
  await waitFor(() => {
    userEvent.click(screen.getByText('Partition #0'));
  });
};

const resetConsumerGroupOffsetsWith = async (
  resetType: string,
  offset: null | number = null
) => {
  userEvent.click(screen.getByLabelText('Reset Type'));
  const options = screen.getAllByText(resetType);
  userEvent.click(options.length > 1 ? options[1] : options[0]);
  userEvent.click(screen.getByText('Select...'));
  await waitFor(() => {
    userEvent.click(screen.getByText('Partition #0'));
  });
  fetchMock.postOnce(
    `/api/clusters/${clusterName}/consumer-groups/${groupId}/offsets`,
    200,
    {
      body: {
        topic: '__amazon_msk_canary',
        resetType,
        partitions: [0],
        partitionsOffsets: [{ partition: 0, offset }],
      },
    }
  );
  userEvent.click(screen.getByText('Submit'));
  await waitFor(() => resetConsumerGroupOffsetsMockCalled());
};

describe('ResetOffsets', () => {
  afterEach(() => {
    fetchMock.reset();
  });

  it('renders progress bar for initial state', async () => {
    fetchMock.getOnce(
      `/api/clusters/${clusterName}/consumer-groups/${groupId}`,
      404
    );
    await waitFor(() => renderComponent());
    expect(screen.getByRole('progressbar')).toBeInTheDocument();
  });

  describe('with consumer group', () => {
    describe('submit handles resetConsumerGroupOffsets', () => {
      beforeEach(async () => {
        const fetchConsumerGroupMock = fetchMock.getOnce(
          `/api/clusters/${clusterName}/consumer-groups/${groupId}`,
          consumerGroupPayload
        );
        await act(() => {
          renderComponent();
        });
        expect(fetchConsumerGroupMock.called()).toBeTruthy();
      });

      it('calls resetConsumerGroupOffsets with EARLIEST', async () => {
        await resetConsumerGroupOffsetsWith('EARLIEST');
      });

      it('calls resetConsumerGroupOffsets with LATEST', async () => {
        await resetConsumerGroupOffsetsWith('LATEST');
      });
      it('calls resetConsumerGroupOffsets with OFFSET', async () => {
        await selectresetTypeAndPartitions('OFFSET');
        fetchMock.postOnce(
          `/api/clusters/${clusterName}/consumer-groups/${groupId}/offsets`,
          200,
          {
            body: {
              topic: '__amazon_msk_canary',
              resetType: 'OFFSET',
              partitions: [0],
              partitionsOffsets: [{ partition: 0, offset: 10 }],
            },
          }
        );
        await waitFor(() => {
          userEvent.click(screen.getAllByLabelText('Partition #0')[1]);
        });
        await waitFor(() => {
          userEvent.keyboard('10');
        });
        userEvent.click(screen.getByText('Submit'));
        await waitFor(() => resetConsumerGroupOffsetsMockCalled());
      });
      it('calls resetConsumerGroupOffsets with TIMESTAMP', async () => {
        await selectresetTypeAndPartitions('TIMESTAMP');
        const resetConsumerGroupOffsetsMock = fetchMock.postOnce(
          `/api/clusters/${clusterName}/consumer-groups/${groupId}/offsets`,
          200,
          {
            body: {
              topic: '__amazon_msk_canary',
              resetType: 'OFFSET',
              partitions: [0],
              partitionsOffsets: [{ partition: 0, offset: 10 }],
            },
          }
        );
        userEvent.click(screen.getByText('Submit'));
        await waitFor(() =>
          expect(
            screen.getByText("This field shouldn't be empty!")
          ).toBeInTheDocument()
        );

        await waitFor(() =>
          expect(
            resetConsumerGroupOffsetsMock.called(
              `/api/clusters/${clusterName}/consumer-groups/${groupId}/offsets`
            )
          ).toBeFalsy()
        );
      });
    });
  });
});
