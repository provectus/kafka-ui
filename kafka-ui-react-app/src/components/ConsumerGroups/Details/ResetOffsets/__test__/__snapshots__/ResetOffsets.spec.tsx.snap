// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetOffsets on initial render matches the snapshot 1`] = `
<body>
  <div>
    <div
      class="section"
    >
      <div
        class="level"
      >
        <div
          class="level-item level-left"
        >
          <nav
            aria-label="breadcrumbs"
            class="breadcrumb"
          >
            <ul>
              <li>
                <a
                  href="/ui/clusters/testCluster/consumer-groups"
                >
                  All Consumer Groups
                </a>
              </li>
              <li>
                <a
                  href="/ui/clusters/testCluster/consumer-groups/testGroup"
                >
                  testGroup
                </a>
              </li>
              <li
                class="is-active"
              >
                <span
                  class=""
                >
                  Reset Offsets
                </span>
              </li>
            </ul>
          </nav>
        </div>
      </div>
      <div
        class="box"
      >
        <form>
          <div
            class="columns"
          >
            <div
              class="column is-one-third"
            >
              <label
                class="label"
                for="topic"
              >
                Topic
              </label>
              <div
                class="select"
              >
                <select
                  id="topic"
                  name="topic"
                >
                  <option
                    value="__amazon_msk_canary"
                  >
                    __amazon_msk_canary
                  </option>
                  <option
                    value="other_topic"
                  >
                    other_topic
                  </option>
                </select>
              </div>
            </div>
            <div
              class="column is-one-third"
            >
              <label
                class="label"
                for="resetType"
              >
                Reset Type
              </label>
              <div
                class="select"
              >
                <select
                  id="resetType"
                  name="resetType"
                >
                  <option
                    value="EARLIEST"
                  >
                    EARLIEST
                  </option>
                  <option
                    value="LATEST"
                  >
                    LATEST
                  </option>
                  <option
                    value="TIMESTAMP"
                  >
                    TIMESTAMP
                  </option>
                  <option
                    value="OFFSET"
                  >
                    OFFSET
                  </option>
                </select>
              </div>
            </div>
            <div
              class="column is-one-third"
            >
              <label
                class="label"
              >
                Partitions
              </label>
              <div
                class="select"
              >
                <div
                  class="rmsc multi-select"
                >
                  <div
                    aria-labelledby="Select partitions"
                    aria-readonly="true"
                    class="dropdown-container"
                    tabindex="0"
                  >
                    <div
                      class="dropdown-heading"
                    >
                      <div
                        class="dropdown-heading-value"
                      >
                        <span
                          class="gray"
                        >
                          Select...
                        </span>
                      </div>
                      <svg
                        class="dropdown-heading-dropdown-arrow gray"
                        fill="none"
                        height="24"
                        stroke="currentColor"
                        stroke-width="2"
                        width="24"
                      >
                        <path
                          d="M6 9L12 15 18 9"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <button
            class="button is-primary"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</body>
`;
